src/main/java/frc/robot/subsystems/SimDrive.java

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import java.util.function.DoubleSupplier;
+
+import com.revrobotics.CANEncoder;
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.EncoderType;
+import com.revrobotics.REVPhysicsSim;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.SparkMaxRelativeEncoder;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+
+import edu.wpi.first.hal.SimDouble;
+import edu.wpi.first.hal.simulation.SimDeviceDataJNI;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.ADXRS450_Gyro;
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.PWM;
+import edu.wpi.first.wpilibj.RobotBase;
+import edu.wpi.first.wpilibj.RobotController;
+import edu.wpi.first.wpilibj.I2C.Port;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.motorcontrol.MotorController;
+import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
+import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;
+import edu.wpi.first.wpilibj.motorcontrol.Talon;
+import edu.wpi.first.wpilibj.simulation.ADXRS450_GyroSim;
+import edu.wpi.first.wpilibj.simulation.DifferentialDrivetrainSim;
+import edu.wpi.first.wpilibj.simulation.EncoderSim;
+import edu.wpi.first.wpilibj.smartdashboard.Field2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+
+public class SimDrive extends SubsystemBase {
+  
+  public SimDrive() {
+
+    SmartDashboard.putData("field", m_field);
+
+    odometer = new DifferentialDriveOdometry(dtSim.getHeading(), leftEncoder.getDistance(), rightEncoder.getDistance());
+    if(RobotBase.isSimulation()){
```

Comment:
it might make more sense to just raise exception. if this isn't simulation, we dont want to be running.

So at the very beginning of this function you can just do:

```java
if (!RobotBase.isSimulation()) {
    raise Exception("BAD, NOT SIMULATION")'
}
```

the value of doing that is that we can then just right away know "this class isn't running if its not simulation".
the problem with that then is that I think we have to add 'throws Exception' or something to the function.  I need to refresh myself on that. 

@doncstokes would probably be able to give some good advice on this.

Reply:
it looks like this is a reasonable thing to do:
```java
   
      DriverStation.reportError("Unable to open trajectory: " + trajectoryJSON, ex.getStackTrace());
   
```

Saw that [here](https://docs.wpilib.org/en/stable/docs/software/pathplanning/pathweaver/integrating-robot-program.html)